torch/csrc/jit/passes/lower_tuples.cpp

```diff
@@ -5,119 +5,35 @@
-    if(TupleTypePtr tt = input->type()->cast<TupleType>()) {
-      JIT_ASSERTM(white_list.count(n->kind()) > 0, "tuple appears in op that does not forward tuples");
-      JIT_ASSERTM(input->node()->kind() == prim::TupleConstruct, "tuple use not matched to tuple construct");
-      for(size_t j = 0; j < tt->elements().size(); ++j) {
-        n->insertInput(i + 1 + j, input->node()->inputs().at(j));
+    if(construct->kind() == prim::TupleConstruct) {
+      for(size_t i = 0; i < n->outputs().size(); ++i) {
+        n->outputs()[i]->replaceAllUsesWith(construct->inputs()[i]);
       }
-      n->removeInput(i);
-      // note: no update to i
-      // since tuples might be nested we need to recursively scan
-      // the new flattened inputs
-    } else {
-      ++i;
+      // op is now dead let dce clean up
+      return;
     }
   }
+
   for(auto b : n->blocks()) {
     LowerTuples(b);
   }
-
-  // flatten the outputs list
-  for(size_t i = 0; i < n->outputs().size();) {
-    Value * output = n->outputs()[i];
-    // (a, b, tup, c) -> (a, b, t0, t1, c)
-    // and:
-    //    tup = (t0, t1)
-    // is placed at the current insertion point
-    if(TupleTypePtr tt = output->type()->cast<TupleType>()) {
-      JIT_ASSERTM(white_list.count(n->kind()) > 0, "tuple appears in op that does not forward tuples");
-      for(size_t j = 0; j < tt->elements().size(); j++) {
-        n->insertOutput(i + 1 + j)->setType(tt->elements()[j]);
-      }
-      auto new_tup = graph.createTuple(n->outputs().slice(i + 1, tt->elements().size()));
-      new_tup->insertBefore(insert_point);
-      insert_point = new_tup;
-      output->replaceAllUsesWith(new_tup->output());
-      n->eraseOutput(i);
-      // note: no update to i to handle nested tuples
-    } else {
-      ++i;
-    }
-  }
 }
 
 static void LowerTuples(Block* block) {
-  // tuples in parameter lists of a block behave exactly the same as
-  // _outputs_ of normal instructions, since the param_node represents the
-  // parameters as outputs, we can handle it by simply visiting the node
-  VisitNode(block->param_node(), *block->nodes().begin());
-  for(auto it = block->nodes().begin(), end = block->nodes().end(); it != end;) {
-    auto n = *it++;
-    VisitNode(n, *it);
-  }
-  // tuples in return lists of blocks behave exactly the same as
-  // _inputs_ of normal instructions, so we can use VisitNode here as well
-  // insert_point is null because it will never be used since return nodes
-  // have no outputs
-  VisitNode(block->return_node(), nullptr);
-}
-
-static void EnsureNoTuples(Block* block) {
-  for (Node* n : block->nodes()) {
-    for (Block* b : n->blocks()) {
-      EnsureNoTuples(b);
-    }
-    for (Value * o : n->outputs()) {
-      JIT_ASSERTM(o->type()->kind() != TypeKind::TupleType,
-                  "Couldn't lower all tuples. This is an error because "
-                  "they're not implemented in the interpreter just yet.");
-    }
+  for(auto n : block->nodes()) {
```

Comment:
Why do we no longer flatten tuples that are inputs or outputs to blocks? This is actually important, because we use it to fix up graphs that take/return tuples into flat arg lists for ONNX.

Reply:
It is not always a good decision to flatten tuples (imagine a huge one, or a 2 element one where one is small and the other huge). We previously had to. I am going to change this so that we leave the old pass to completely flatten for onnx export (though disturbingly the onnx tests do not fail with the code as it), while we only typically run the new very simple pass.